lib/internal/modules/esm/import_map.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+const { isURL, URL } = require('internal/url');
+const { ObjectEntries, ObjectKeys, SafeMap, ArrayIsArray } = primordials;
+const { codes: { ERR_INVALID_IMPORT_MAP } } = require('internal/errors');
+
+class ImportMap {
+  #baseURL;
+  imports = new SafeMap();
+  scopes = new SafeMap();
+
+  constructor(raw, baseURL) {
+    this.#baseURL = baseURL;
+    processImportMap(this, this.#baseURL, raw);
+  }
+
+  get baseURL() {
+    return this.#baseURL;
+  }
+
+  resolve(specifier, parentURL = this.baseURL) {
+    // Process scopes
+    for (const { 0: prefix, 1: mapping } of this.scopes) {
+      let mappedSpecifier = mapping.get(specifier);
+      if (parentURL.pathname.startsWith(prefix.pathname) && mappedSpecifier) {
+        if (!isURL(mappedSpecifier)) {
+          mappedSpecifier = new URL(mappedSpecifier, this.baseURL);
+          mapping.set(specifier, mappedSpecifier);
+        }
+        specifier = mappedSpecifier;
+        break;
+      }
+    }
+
+    let spec = specifier;
+    if (isURL(specifier)) {
+      spec = specifier.pathname;
+    }
+    let importMapping = this.imports.get(spec);
+    if (importMapping) {
+      if (!isURL(importMapping)) {
+        importMapping = new URL(importMapping, this.baseURL);
+        this.imports.set(spec, importMapping);
+      }
+      return importMapping;
+    }
+
+    return specifier;
+  }
+}
+
+function processImportMap(importMap, baseURL, raw) {
+  // Validation and normalization
+  if (typeof raw.imports !== 'object' || ArrayIsArray(raw.imports)) {
+    throw new ERR_INVALID_IMPORT_MAP('top level key "imports" is required and must be a plain object');
+  }
+  if (typeof raw.scopes !== 'object' || ArrayIsArray(raw.scopes)) {
+    throw new ERR_INVALID_IMPORT_MAP('top level key "scopes" is required and must be a plain object');
+  }
+
+  // Normalize imports
+  for (const { 0: specifier, 1: mapping } of ObjectEntries(raw.imports)) {
+    if (!specifier || typeof specifier !== 'string') {
+      throw new ERR_INVALID_IMPORT_MAP('module specifier keys must be non-empty strings');
+    }
+    if (!mapping || typeof mapping !== 'string') {
+      throw new ERR_INVALID_IMPORT_MAP('module specifier values must be non-empty strings');
+    }
+    if (specifier.endsWith('/') && !mapping.endsWith('/')) {
+      throw new ERR_INVALID_IMPORT_MAP('module specifier values for keys ending with / must also end with /');
```

Comment:
```suggestion
      throw new ERR_INVALID_IMPORT_MAP('module specifier keys ending with "/" must have values that end with "/"');
```

Reply:
> We have generic helpers like validateString, if there isn’t a validatePlainObject it might be worth creating one.

Not sure why github did not give me a comment box to directly reply to that comment, but here it is.  I think that would be good idea and I can add them in this PR if they did not exist. Where would they live if I wanted to go looking for methods like this?`